// reason-apollo/src/ReasonApollo.bs.js
// Generated by BUCKLESCRIPT VERSION 5.0.3, PLEASE EDIT WITH CARE

import * as Caml_option from "bs-platform/lib/es6/caml_option.js";
import * as ApolloClient from "apollo-client";
import * as ReasonApolloQuery from "./graphql-types/ReasonApolloQuery.bs.js";
import * as ReasonApolloMutation from "./graphql-types/ReasonApolloMutation.bs.js";
import * as ReasonApolloSubscription from "./graphql-types/ReasonApolloSubscription.bs.js";

function createApolloClient(link, cache, ssrMode, ssrForceFetchDelay, connectToDevTools, queryDeduplication, param) {
  var tmp = {
    link: link,
    cache: cache
  };
  if (ssrMode !== undefined) {
    tmp.ssrMode = Caml_option.valFromOption(ssrMode);
  }
  if (ssrForceFetchDelay !== undefined) {
    tmp.ssrForceFetchDelay = Caml_option.valFromOption(ssrForceFetchDelay);
  }
  if (connectToDevTools !== undefined) {
    tmp.connectToDevTools = Caml_option.valFromOption(connectToDevTools);
  }
  if (queryDeduplication !== undefined) {
    tmp.queryDeduplication = Caml_option.valFromOption(queryDeduplication);
  }
  return new ApolloClient.ApolloClient(tmp);
}

var CreateQuery = ReasonApolloQuery.Make;

var CreateMutation = ReasonApolloMutation.Make;

var CreateSubscription = ReasonApolloSubscription.Make;

var Provider = 0;

var Consumer = 0;

export {
  createApolloClient ,
  CreateQuery ,
  CreateMutation ,
  CreateSubscription ,
  Provider ,
  Consumer ,
  
}
/* apollo-client Not a pure module */
